p2p/src/tests/bad_time_diff.rs

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{sync::Arc, time::Duration};
+
+use common::primitives::{time::Time, user_agent::mintlayer_core_user_agent};
+use p2p_test_utils::{run_with_timeout, P2pBasicTestTimeGetter};
+use test_utils::assert_matches;
+
+use crate::{
+    config::P2pConfig,
+    disconnection_reason::DisconnectionReason,
+    message::WillDisconnectMessage,
+    net::default_backend::{
+        transport::{BufferedTranscoder, TransportListener, TransportSocket},
+        types::{HandshakeMessage, Message, P2pTimestamp},
+    },
+    protocol::SupportedProtocolVersion,
+    testing_utils::{
+        TestTransportChannel, TestTransportMaker, TestTransportNoise, TestTransportTcp,
+    },
+    tests::helpers::TestNode,
+};
+
+// Check that a handshake is rejected if the time difference between the peers is too big.
+// Also check that the WillDisconnect message is sent in this case.
+
+// V3+ is needed for the WillDisconnect message to be sent.
+const TEST_PROTOCOL_VERSION: SupportedProtocolVersion = SupportedProtocolVersion::V3;
```

Comment:
Are there tests for how V3 will act when meeting V2?

Reply:
There are "generic" tests that check what happens when different protocol versions meet each other (the lowest version is selected). I don't think we need additional tests for this.

There was one thing I missed though - to check that `WillDisconnect` won't be sent for V2. So I've pushed an additional commit where I check this (by running the `bad_time_diff` and `same_handshake_nonce` tests for V2 as well.)